File: OMSWeb/oms-track-monitor/src/segment/utils/segment.ts
import { ISegmentPart } from "../../legacies/models/track.model";
import { encodeCommandsToD } from "../../utils/svg/pathSegment";
import { makeD } from "../segments";
import { Segment } from "../types/Segment";

function makeSegmentsFromParts(parts: ISegmentPart[]): Segment[] {
  const map = new Map<number, ISegmentPart[]>()
  parts.forEach(segmentPart => {
    const parts = map.get(segmentPart.id) ?? []
    map.set(segmentPart.id, [...parts, segmentPart])
  })

  const array = Array<Segment>()
  map.forEach((parts, id) => {
    if (parts[0]) {
      const sample = parts[0]
      const sortedParts = parts.sort((a, b) => a.segpartId - b.segpartId)
      const pathCommands = makeD(
        sample.startPoint,
        sample.endPoint,
        // @ts-ignore
        sortedParts,
        sample.length
      )
      array.push({
        id: id,
        logicalId: sample.logicalId,
        physicalId: sample.physicalId,

        startPoint: sample.startPoint,
        endPoint: sample.endPoint,
        length: sample.length,
        speed: sample.speed,

        disabled: false,

        // @ts-ignore
        parts: sortedParts,
        pathCommands,
        d: encodeCommandsToD(pathCommands)
      })
    }
  })

  return array
}

export { makeSegmentsFromParts }

File: OMSWeb/Models/Tracks/Segment.cs
namespace OMSWeb.Models.Tracks
{
  public interface ISegment
  {
    int Id { get; set; }
    string PhysicalId { get; set; }
    string LogicalId { get; set; }
    int StartPoint { get; set; }
    int EndPoint { get; set; }
    float Speed { get; set; }
    float Length { get; set; }
  }
  public interface ISegmentPart
  {
    int Id { get; set; }
    string Type { get; set; }
    string Direction { get; set; }
    string Location { get; set; }

    // int? segmentId { get; set; }
  }
  public class Segment : ISegment
  {
    public int Id { get; set; }
    public string PhysicalId { get; set; }
    public string LogicalId { get; set; }
    public int StartPoint { get; set; }
    public int EndPoint { get; set; }
    public float Speed { get; set; }
    public float Length { get; set; }
    
    public int SteerDir { get; set; }
    public int SpeedRatio { get; set; }
    public int Oblow { get; set; }
    public int Obhigh { get; set; }
  }
  public class SegmentWithPart : Segment, ISegmentPart
  {
    public string Type { get; set; }
    public string Direction { get; set; }
    public string Location { get; set; }

    public int SegpartId { get; set; }

  }
}